src/public/target/sections/Safety/Section.js

```diff
@@ -1,221 +1,198 @@
-          </React.Fragment>
-        );
-      });
-    },
+    renderCell: ({ activationEffects }) =>
+      listTermsByGroup(activationEffects).map(({ groupKey, terms }) => (
+        <React.Fragment key={groupKey}>
+          <Typography variant="subtitle2">
+            {_.capitalize(_.lowerCase(groupKey))}
+          </Typography>
+          <ul>
+            {terms.map(term => (
+              <li key={`${groupKey}:${term}`}>{term}</li>
+            ))}
+          </ul>
+        </React.Fragment>
+      )),
   },
   {
     id: 'inhibition_effects',
     label: 'Antagonism or inhibition effects',
-    renderCell: ({ inhibition_effects: inhibitionEffects }) => {
-      return Object.keys(inhibitionEffects).map(key => {
-        return (
-          <React.Fragment key={key}>
-            <Typography variant="subtitle2">{capitalize(key)}</Typography>
-            <ul>
-              {inhibitionEffects[key].map((effect, i) => (
-                <li key={i}>{effect.mapped_term || effect.term_in_paper}</li>
-              ))}
-            </ul>
-          </React.Fragment>
-        );
-      });
-    },
+    renderCell: ({ inhibitionEffects }) =>
+      listTermsByGroup(inhibitionEffects).map(({ groupKey, terms }) => (
+        <React.Fragment key={groupKey}>
+          <Typography variant="subtitle2">
+            {_.capitalize(_.lowerCase(groupKey))}
+          </Typography>
+          <ul>
+            {terms.map(term => (
+              <li key={`${groupKey}:${term}`}>{term}</li>
+            ))}
+          </ul>
+        </React.Fragment>
+      )),
   },
   {
     id: 'references',
     label: 'Publications',
-    renderCell: ({ references }) => {
-      return references.map((reference, i) => {
-        const isHecatos = reference.ref_link.indexOf('hecatos') !== -1;
-        return (
-          <React.Fragment key={i}>
-            <Link external to={reference.ref_link}>
-              <Tooltip
-                title={
-                  isHecatos
-                    ? "HeCaToS Deliverable D01.5 (2015) funded by 'EU 7th Framework Programme (HEALTH-F4-2013-602156).'"
-                    : ''
-                }
-                placement="top"
-              >
-                <span>{reference.ref_label}</span>
-              </Tooltip>
-            </Link>{' '}
-          </React.Fragment>
-        );
-      });
-    },
+    renderCell: ({ references }) => (
+      <ReferencesCell>{references}</ReferencesCell>
+    ),
   },
 ];
 
 const riskColumns = [
   {
     id: 'organs_systems_affected',
     label: 'Main organs & systems affected',
-    renderCell: ({ organs_systems_affected: organs }) => {
-      return (
-        <ul>
-          {organs.map((organ, i) => (
-            <li key={i}>{organ.mapped_term || organ.term_in_paper}</li>
+    renderCell: ({ organsSystemsAffected }) => (
+      <ul>
+        {organsSystemsAffected
+          .filter(
+            organ => organ.code && (organ.mappedTerm || organ.termInPaper)
```

Comment:
I identified four bits of data transformation logic that are pushed into the frontend by the GraphQL schema:

1. All fields that use the fragment `termOptions` in the GraphQL query I wrote (all requested `SafetyCode`-typed fields except the `target → safety → adverseEffects → organsSystemsAffected` I commented on [above](https://github.com/opentargets/platform-app/pull/146/files#r425934618)) do so because the view is interested in either the `mappedTerm` field _or_ the `termInPaper` field, whichever is available. This logic could be moved to the server side by adding a field that contains this preferred term.
2. The two `SafetyReference` fields in the query request both the `pubmedId` field and the `refLink` field, again because I either format the `pubmedId` into a link or fall back on the `refLink`; I ported this logic from the Angular webapp. This logic could be moved to the server side by adding a field that contains this preferred link.
3. No part of the section is interested in objects for which either of the two combined fields I mentioned are `null`—I had to add logic to filter out objects that the server might send in which they were.
4. I believe the frontend will always want to request all members of `target → safety → adverseEffects → activationEffects` and `target → safety → adverseEffects → inhibitionEffects` and convert them to an array of `{groupKey, term[]}` pairs. This transformation logic (the function [`listTermsByGroup()`](https://github.com/opentargets/platform-app/pull/146/files#diff-840a84ad4f200ab6b91e08506a4a9c20R186) below) could be moved to the server side by adding two such array fields to the `AdverseEffects` type.

If these four bits of logic survive review, it may be interesting to track them as (GitHub) issues, depending on the software-architectural direction the project is aiming for.

Reply:
I prototyped how much cleaner the presentation code would look [here](https://github.com/thehyve/opentargets_platform_app/commit/port-target-safety-query-container-component#diff-84b38b23b276f8f908a2c491ecc1bf35), as the code duplication was bugging me over the weekend :smile: